### PR TITLE
Send email notification with build status, repository link and cause of build

### PR DESCRIPTION
**Pull request by navidshaikh**
This PR implements sending email notifications to users with

Status of build
Repository link (which can be opened in browser in near future, where all the build results can be seen)
Cause of the build
Changes

Adds a new stage Notify user build pipeline
Added new env vars in seed-job/template.yml
Added ccp/notifications module
Added library modules for
email utilities (any module can use it and send emails)
OpenShift utility to get Jenkins authorization token from jenkins secret using oc
OpenShift utility to get Jenkins build information
Added unittests for notifications and library modules

### CODE REVIEW

File: ccp/index_reader.py [hunk 1]
```diff
@@ -265,11 +265,15 @@ class BuildConfigManager(object):
     by pipeline service
     """
 
-    def __init__(self, registry_url, namespace, from_address, smtp_server):
+    def __init__(self, registry_url, namespace, from_address, smtp_server,
+                 ccp_openshift_slave_image, notify_cc_emails):
```
Comment (navidshaikh) [inline ccp/index_reader.py]: parameters which needed to be passed from build template to template

File: ccp/index_reader.py [hunk 2]
```diff
@@ -265,11 +265,15 @@ class BuildConfigManager(object):
...
         self.registry_url = registry_url
         self.namespace = namespace
         self.from_address = from_address
         self.smtp_server = smtp_server
+        self.ccp_openshift_slave_image = ccp_openshift_slave_image
+        self.notify_cc_emails = notify_cc_emails
```
Comment (navidshaikh) [inline ccp/index_reader.py]: parameters which needed to be passed from build template to template

File: ccp/index_reader.py [hunk 3]
```diff
@@ -648,7 +661,7 @@ def batch_process_projects(self,
 
 
 if __name__ == "__main__":
-    if len(sys.argv) != 9:
+    if len(sys.argv) != 11:
```
Comment (navidshaikh) [inline ccp/index_reader.py]: have added the card for this to convert to argument processor already, will be part of separate PR.

**Comment by navidshaikh** [comment]
#dotests

**Comment by navidshaikh** [comment]
#dotests-unittests

File: ccp/lib/command.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    :param shell: Whether to run raw shell commands with '|' and redirections
+    :type cmd: str
+    :type shell: boolean
+
+    :return: Command's stdout and stderr
+    :rtype: tuple
+    :raises: subprocess.CalledProcessError
+    """
+    p = subprocess.Popen(cmd, shell=shell)
+    return p.communicate()
```
Comment (navidshaikh) [inline ccp/lib/command.py]: This is for logging the stderr errors.

**Comment by navidshaikh** [comment]
The CI is failing with
```
Error while deploying the service in CICO
```

File: ccp/lib/command.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    :rtype: str
+    :raises: subprocess.CalledProcessError
+    """
+    if shell:
+        return subprocess.check_output(cmd, shell=True)
+    else:
+        return subprocess.check_output(cmd.split(), shell=False)
+
+
+def run_cmd2(cmd, shell=False):
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/lib/command.py]: Even the standard library methods also follows this convention when the behavior is slightly different but use case is same
```
In [4]: from shutil import copy

In [5]: from shutil import copy2
```

File: ccp/lib/email.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    def email(self,
+              smtp_server, sub, body,
+              from_add, to_adds, cc_adds=[]):
+        """
+        Send email using given details
+        smtp_server: URL of smtp server
+        sub: Subject of email
+        body: Body of email
+        from_add: From address
+        to_adds: A list of to addresses
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/lib/email.py]: Yes - The back-end has provision for it. We need to think later how we want to implement this index if we plan to add this.

File: ccp/lib/openshift.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    # {{ --> { i.e. double curly brackets into single
+
+    print ("Get Jenkins service account token\n{}".format(command))
+    # run the oc command
+    token = run_cmd(command, shell=True)
+    # strip any extra characters while reading stdout
+    return token.strip()
+
+
+class BuildInfo(object):
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/lib/openshift.py]: Thought about it while naming it, however the build is trigerred from OpenShift and executed at Jenkins.

In future after prod movement, once we get the clients and processors in, we will take this part of the code aligned together.

File: ccp/lib/openshift.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+        print ("Opening URL to details of the build\n{}".format(url))
+
+        try:
+            token = get_jenkins_access_token(self.service_account)
+            response = self.get_url(url, token)
+            response = json.loads(response.read())
+        except Exception as e:
+            print ("Error opening URL {}".format(e))
+            raise(e)
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/lib/openshift.py]: Yes, with some detail about which URL opening failed.

File: ccp/notifications/base.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    classes.
+    """
+
+    def __init__(self):
+        self.build_success_subj = \
+            "[registry.centos.org] SUCCESS: Container build {}"
+        self.build_failure_subj = \
+            "[registry.centos.org] FAILED: Container build {}"
+
+        self.build_success_body = """\
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/notifications/base.py]: Yes, the modules/formatting/etc we dont know in programming, we need to look up for it. Its a formatting syntax with right spacing defined.

File: ccp/notifications/notify.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        body = body + "\n\n" + self.email_footer
+
+        return body
+
+    def notify(self,
+               status, namespace, jenkins_url,
+               image_name, build_number, pipeline_name):
+        """
+        Get notifications info and sends email to user
+        Prints success message upon successful email delivery, else
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/notifications/notify.py]: take a look here https://github.com/CentOS/container-pipeline-service/pull/616/files#diff-bf45a09055b56a3a1ec7a5e822e96358R139

File: seed-job/template.yaml [hunk 11]
```diff
@@ -132,14 +132,16 @@ objects:
...
-                        if (success==true) {
-                            sh "echo -e 'Build request for ${image_name_with_registry} was successful' | mail -r ${FROM_ADDRESS} -S smtp=${SMTP_SERVER} -s 'SUCCESS: Container build ${image_name_with_registry} is complete' ${NOTIFY_EMAIL} "
-                        } else {
-                            sh "echo -e 'Build request for ${image_name_with_registry} was unsuccessful' | mail -r ${FROM_ADDRESS} -S smtp=${SMTP_SERVER} -s 'FAILED: Container build ${image_name_with_registry} has failed' ${NOTIFY_EMAIL} "
-
-                        }
+                      stage("Notify user"){
+                          sh "mkfifo /var/spool/postfix/public/pickup && postfix start"
+                          if (success==true) {
+                            sh "PYTHONPATH=${PIPELINE_REPO_DIR} python ${PIPELINE_REPO_DIR}/ccp/notifications/notify.py success cccp `oc get route jenkins -o template --template={{.spec.host}}` ${image_name} ${BUILD_NUMBER} ${PIPELINE_NAME}"
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply seed-job/template.yaml]: Take a look at the card created in earlier PR review. Its already updated to scope out this concern.

[the remark this answers is not in the document]
Comment (navidshaikh) [reply seed-job/template.yaml]:
```
[Pipeline] sh
++ oc project -q
error: no project has been set
```

so we can't list project from jenkins slave pod container.

I remember removing this hard coding  and trying to fix it with this approach only, however it was left off behind due to other PR's work.

A cleaner solution would be to carry the `NAMESPACE` parameter passed in buildtemplate.yaml to template.yaml.

File: ccp/notifications/base.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+class BaseNotify(object):
+    """
+    BaseNotify class has related common methods and
+    initialization for build and weekly scan notification
+    classes.
+    """
+
+    def __init__(self):
+        self.build_success_subj = \
+            "[registry.centos.org] SUCCESS: Container build {}"
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/notifications/base.py]: Make sense.

We need to make sure that when we deploy things in prod, we set the REGISTRY_URL to `registry.centos.org`.

File: ccp/notifications/base.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+{0:<30}{1}
+{2:<30}{3}"""
+
+        self.email_footer = """\
+--
+Do you have a query?
+Talk to Pipeline team on #centos-devel at freenode
+CentOS Community Container Pipeline Service
+https://wiki.centos.org/ContainerPipeline
+https://github.com/centos/container-index
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/notifications/base.py]: Make sense to me, updating it to
```
--
Do you have a query?
Talk to CentOS Container Pipeline team on #centos-devel at freenode
https://wiki.centos.org/ContainerPipeline
```

File: ccp/notifications/base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/notifications/base.py]: Added a comment anyway.

File: ccp/lib/openshift.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/lib/openshift.py]: For eg, the utility is working with service account token, namespace, pipeline name.

File: ccp/lib/openshift.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+
+import json
+import ssl
+import sys
+import urllib2
+
+from ccp.lib.command import run_cmd
+
+
+def get_jenkins_access_token(service_account="sa/jenkins"):
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/lib/openshift.py]: yes, I will need to define the decorator in a separate module in library and re-use at other places.

File: ccp/lib/openshift.py [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+        Jenkins and specifying the required_fields expected while
+        retrieving build information.
+        """
+        self.service_account = service_account
+        # given the list of fields,
+        self.required_fields = required_fields
+
+    def get_url(self, url, token, context=None):
+        """
+        Gets the given URL and uses given token for authorization
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/lib/openshift.py]: It just opens the URL and return the whole response by loading in JSON. Its not parsed.

Updating the docstring to specify ^

File: ccp/lib/openshift.py [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+                                "hudson.model.Cause$UpstreamCause":
+                            # process upstream project name
+                            # eg: cccp/cccp-test-anomaly-latest
+                            ups_proj = cause["upstreamProject"].split(
+                                "/")[-1].split("-")
+                            # form appid/jobid:desiredtag
+                            ups_proj = "/".join(
+                                ups_proj[1:-1]) + ":" + ups_proj[-1]
+
+                            cause_of_build = ("Upstream/parent container {} is"
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply ccp/lib/openshift.py]: To be explicit, using `parent container image [..]`

File: seed-job/template.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (dharmit) [reply seed-job/template.yaml]: Yes, as we do before triggering the seed-job. `NAMESPACE` is already available in `buildtemplate.yaml`.

File: ccp/lib/openshift.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dharmit) [reply ccp/lib/openshift.py]: I think neither of us is wrong. It's just that `JenkinsBuildInfo` to me sounds like something very explicitly Jenkins. Which is what we're referring to in this case to get the build info.

Marking this resolved.

File: ccp/index_reader.py (hunk 3 above)
Comment (mohammedzee1000) [reply ccp/index_reader.py]: This is needed. Why we even went argv route here is beyond me

File: ccp/lib/_print.py [hunk 17]
```diff
@@ -0,0 +1,15 @@
+"""
+This library util module has _print method defined
+which can be used to print messages on stdout realtime
+and not being blocked on stdout buffer for showing the messages.
+"""
+
+import sys
+
+
+def _print(msg):
```
Comment (mohammedzee1000) [inline ccp/lib/_print.py]: Maybe add verbose parameter here which is default True, so it can be used with those cases as well

**Comment by dharmit** [comment]
LGTM

File: ccp/lib/command.py [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    :param cmd: Command to run
+    :param shell: Whether to run raw shell commands with '|' and redirections
+    :type cmd: str
+    :type shell: boolean
+
+    :return: Command's stdout and stderr
+    :rtype: tuple
+    :raises: subprocess.CalledProcessError
+    """
+    p = subprocess.Popen(cmd, shell=shell)
```
Comment (mohammedzee1000) [inline ccp/lib/command.py]: This should be the default way, not check_call

File: Dockerfiles/ccp-openshift-slave/Dockerfile [hunk 19]
```diff
@@ -1,6 +1,8 @@
 FROM registry.centos.org/openshift/jenkins-slave-base-centos7
 
-RUN yum install -y PyYAML docker mailx epel-release && \
+RUN yum install -y PyYAML docker mailx postfix epel-release && \
     yum install -y npm && \
     yum clean all && \
     npm install -g dockerfile_lint
+
+ADD ./ /opt/ccp-openshift/
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply Dockerfiles/ccp-openshift-slave/Dockerfile]: Yes, this container should be built from root dir of git repo (build context) while referencing the Dockerfile using -f to the abs path of the slave Dockerfile.

File: ccp/index_reader.py (hunk 1 above)
Comment (navidshaikh) [reply ccp/index_reader.py]: Guess - I'll use the jenkins REST APIs to fetch these details in IndexReader later as we did in notifications module. ;-)

Dont wanted to overload this PR.

File: ccp/index_reader.py (hunk 3 above)
Comment (navidshaikh) [reply ccp/index_reader.py]: It started with few arguments, and kept growing. Again, if we start using the Jenkins REST APIs in index_reader, the number of arguments to be passed would decrease as we can grab that via the APIs itself. Even if they are less, arg processor is needed.

File: ccp/lib/_print.py (hunk 17 above)
Comment (navidshaikh) [reply ccp/lib/_print.py]: I am planning to add logging mechanism here.

If set to a file will export to file or to console, will add all those options then.

File: ccp/lib/command.py (hunk 18 above)
Comment (navidshaikh) [reply ccp/lib/command.py]: where are we using check_call ? @mohammedzee1000

File: ccp/lib/openshift.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (dharmit) [reply ccp/lib/openshift.py]: :+1:

**Comment by navidshaikh** [comment]
#dotests

**Comment by navidshaikh** [comment]
#dotests-unittests

**Comment by navidshaikh** [comment]
#dotests

**Comment by navidshaikh** [comment]
#dotests-unittests

**Comment by navidshaikh** [comment]
#dotests
